tests/unit/test_postgresql_provider.py

```diff
@@ -0,0 +1,266 @@
+            self.assertFalse(isinstance(self.harness.model.unit.status, BlockedStatus))
+
+            # BlockedStatus due to a PostgreSQLCreateUserError.
+            self.request_database()
+            self.assertTrue(isinstance(self.harness.model.unit.status, BlockedStatus))
+            # No data is set in the databag by the database.
+            self.assertEqual(
+                self.harness.get_relation_data(self.rel_id, self.app),
+                {
+                    "data": f'{{"database": "{DATABASE}", "extra-user-roles": "{EXTRA_USER_ROLES}"}}',
+                },
+            )
+
+            # BlockedStatus due to a PostgreSQLCreateDatabaseError.
+            self.request_database()
+            self.assertTrue(isinstance(self.harness.model.unit.status, BlockedStatus))
+            # No data is set in the databag by the database.
+            self.assertEqual(
+                self.harness.get_relation_data(self.rel_id, self.app),
+                {
+                    "data": f'{{"database": "{DATABASE}", "extra-user-roles": "{EXTRA_USER_ROLES}"}}',
+                },
+            )
+
+            # BlockedStatus due to a PostgreSQLGetPostgreSQLVersionError.
+            self.request_database()
+            self.assertTrue(isinstance(self.harness.model.unit.status, BlockedStatus))
+
+    @patch.object(EventBase, "defer")
+    @patch("charm.Patroni.member_started", new_callable=PropertyMock)
+    def test_on_relation_broken(self, _member_started, _defer):
+        with patch.object(PostgresqlOperatorCharm, "postgresql", Mock()) as postgresql_mock:
+            # Set some side effects to test multiple situations.
+            _member_started.side_effect = [False, True, True]
+            postgresql_mock.delete_user = PropertyMock(
+                side_effect=[None, PostgreSQLDeleteUserError]
+            )
+
+            # Break the relation before the database is ready.
+            self.harness.remove_relation(self.rel_id)
+            _defer.assert_called_once()
+
+            # Assert that the correct calls were made after a relation broken event.
+            self.rel_id = self.harness.add_relation(RELATION_NAME, "application")
+            self.harness.remove_relation(self.rel_id)
+            user = f"relation_id_{self.rel_id}"
+            postgresql_mock.delete_user.assert_called_once_with(user)
+
+            # Test a failed user deletion.
+            self.rel_id = self.harness.add_relation(RELATION_NAME, "application")
+            self.harness.remove_relation(self.rel_id)
+            self.assertTrue(isinstance(self.harness.model.unit.status, BlockedStatus))
+
+    @patch(
+        "charm.PostgresqlOperatorCharm.primary_endpoint",
+        new_callable=PropertyMock(return_value="1.1.1.1"),
+    )
+    @patch(
+        "charm.PostgresqlOperatorCharm.members_ips",
+        new_callable=PropertyMock,
+    )
+    def test_update_endpoints_with_event(self, _members_ips, _primary_endpoint):
```

Comment:
These tests are all really nice, I want to go back and redo the pgbouncer ones this way